### DISCUSSION TITLE
How do you know the score range of each recognizer?

### DISCUSSION THREAD

**Discussion by lifepillar** [microsoft/presidio#1170] [Q&A]
Is there a way to query what scores a recognizer can produce with and without context enhancement? Suppose I want to write my own specialized recognizer, and I want it to beat a given Presidio's recognizer *R*, except when *R*'s score is boosted by context. For that, I need to know what scores *R* assigns (base score and, at least, the minimal enhanced score). Is there a way to get such info other than browsing through the code?

**Comment by omri374** [comment]
Each recognizer has a different way of assigning scores. For regex recognizers (`PatternRecognizer`) the score is determined by the Pattern score, the context score, and a validation score.
For the pattern score, see the list of patterns used by the recognizer, for example:
https://github.com/microsoft/presidio/blob/d8541e973349ceefccec926a01992dd554de6123/presidio-analyzer/presidio_analyzer/predefined_recognizers/uk_nhs_recognizer.py#L19

For context score, the value is determined by the `ContextAwareEnhancer`. In the default case, the value is set by `context_similarity_factor` which is 0.35. 
https://github.com/microsoft/presidio/blob/d8541e973349ceefccec926a01992dd554de6123/presidio-analyzer/presidio_analyzer/context_aware_enhancers/lemma_context_aware_enhancer.py#L138

For checksum validation, the score is determined by `EntityRecognizer.MAX_SCORE`, which is set to 1.0 by default.
All these could be changed for a specific implementation.

To get more insights for every decision, you can return the `decision_process` output which will contain the score before and after each step :https://microsoft.github.io/presidio/analyzer/decision_process/
